examples/step-67/step-67.cc

```diff
@@ -0,0 +1,2305 @@
+  // have moved the decision out of the quadrature point loop. However, the
+  // loss of efficiency is hardly noticeable, so we opt for the simpler code
+  // here. Also note that the final `else` clause will catch the case when
+  // some part of the boundary was not assigned any boundary condition via
+  // `EulerOperator::set_..._boundary(...)`.
+  template <int dim, int degree, int n_points_1d>
+  void EulerOperator<dim, degree, n_points_1d>::local_apply_boundary_face(
+    const MatrixFree<dim, Number> &,
+    LinearAlgebra::distributed::Vector<Number> &      dst,
+    const LinearAlgebra::distributed::Vector<Number> &src,
+    const std::pair<unsigned int, unsigned int> &     face_range) const
+  {
+    FEFaceEvaluation<dim, degree, n_points_1d, dim + 2, Number> phi(data, true);
+
+    for (unsigned int face = face_range.first; face < face_range.second; face++)
+      {
+        phi.reinit(face);
+        phi.gather_evaluate(src, true, false);
+
+        for (unsigned int q = 0; q < phi.n_q_points; ++q)
+          {
+            const auto w_m    = phi.get_value(q);
+            const auto normal = phi.get_normal_vector(q);
+
+            auto rho_u_dot_n = w_m[1] * normal[0];
+            for (unsigned int d = 1; d < dim; ++d)
+              rho_u_dot_n += w_m[1 + d] * normal[d];
+
+            bool at_outflow = false;
+
+            Tensor<1, dim + 2, VectorizedArray<Number>> w_p;
+            const auto boundary_id = data.get_boundary_id(face);
+            if (wall_boundaries.find(boundary_id) != wall_boundaries.end())
+              {
+                w_p[0] = w_m[0];
+                for (unsigned int d = 0; d < dim; ++d)
+                  w_p[d + 1] = w_m[d + 1] - 2. * rho_u_dot_n * normal[d];
+                w_p[dim + 1] = w_m[dim + 1];
+              }
+            else if (inflow_boundaries.find(boundary_id) !=
+                     inflow_boundaries.end())
+              w_p =
+                evaluate_function(*inflow_boundaries.find(boundary_id)->second,
+                                  phi.quadrature_point(q));
+            else if (subsonic_outflow_boundaries.find(boundary_id) !=
+                     subsonic_outflow_boundaries.end())
+              {
+                w_p          = w_m;
+                w_p[dim + 1] = evaluate_function(
+                  *subsonic_outflow_boundaries.find(boundary_id)->second,
+                  phi.quadrature_point(q),
+                  dim + 1);
+                at_outflow = true;
+              }
+            else if (supersonic_outflow_boundaries.find(boundary_id) !=
+                     supersonic_outflow_boundaries.end())
+              {
+                w_p        = w_m;
+                at_outflow = true;
+              }
+            else
+              AssertThrow(false,
+                          ExcMessage("Unknown boundary id, did "
+                                     "you set a boundary condition?"));
```

Comment:
I kind of have to, because I otherwise force the compiler to inline the numerical flux 4 times, and I would believe that duplicating the code is worse for the instruction cache. Thinking about it, I wonder I should remove the `always_inline` decoration from the numerical flux (but not the other places), so I could share the code among a few places. Well, I think it's better not to, as the compiler will likely mess things up.